src/libUtils/Bech32.cpp

```diff
@@ -0,0 +1,342 @@
+/*
+ * Copyright (C) 2021 Zilliqa
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include <algorithm>
+#include <boost/algorithm/string.hpp>
+#include <boost/utility/string_view.hpp>
+#include <iostream>
+#include <regex>
+
+#include "Bech32.h"
+#include "libServer/AddressChecksum.h"
+
+const std::string Bech32::CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";
+
+const char Bech32::CHARSET_REV[128] = {
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 15, -1, 10, 17, 21, 20, 26, 30, 7,
+    5,  -1, -1, -1, -1, -1, -1, -1, 29, -1, 24, 13, 25, 9,  8,  23, -1, 18, 22,
+    31, 27, 19, -1, 1,  0,  3,  16, 11, 28, 12, 14, 6,  4,  2,  -1, -1, -1, -1,
+    -1, -1, 29, -1, 24, 13, 25, 9,  8,  23, -1, 18, 22, 31, 27, 19, -1, 1,  0,
+    3,  16, 11, 28, 12, 14, 6,  4,  2,  -1, -1, -1, -1, -1};
+
+const std::string Bech32::ZIL_HRP = "zil";
+
+const char Bech32::HEX_ARRAY[16] = {'0', '1', '2', '3', '4', '5', '6', '7',
+                                    '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
+
+const std::regex Bech32::BECH32_ZIL_REGEX_FORMAT(
+    "^zil1[qpzry9x8gf2tvdw0s3jn54khce6mua7l]{38}");
+const std::regex Bech32::BASE16_ZIL_REGEX_FORMAT("^(0x)?[0-9a-fA-F]{40}");
+
+int Bech32::hexCharToInt(char hexChar) {
```

Comment:
Using streams to convert individual characters might be inefficient. Something like this should suffice:
```
int i = -1;
if (hexChar >= '0' && hexChar <= '9') {
  i = hexChar - '0';
} else if (hexChar >= 'a' && hexChar <= 'f') {
  i = 10 + hexChar - 'a';
}
return i;
```

Reply:
Additionally, if you also check for `(hexChar >= 'A' && hexChar <= 'F')` then `hexStringToCharVec` won't need to call `boost::to_lower_copy` anymore